### PR TITLE
svcmain: use hexops/cmder for CLI arg parsing

### PR DESCRIPTION
**Pull request by slimsag**
@keegancsmith this is an alternative to your PR https://github.com/sourcegraph/sourcegraph/pull/49518 which uses urfave/cli.

I thought I'd whip this up real quick, in case folks like this approach better. [hexops/cmder](https://github.com/hexops/cmder) (by yours truly) I think is a bit nicer than urface/cli because:

1. It's lighter weight; compare https://github.com/urfave/cli with the ~100 LOC file that is cmder: https://github.com/hexops/cmder/blob/main/cmder.go - just less magical overall.
2. It is directly modeled after what the `go` tool does internally; uses the Go `flag` package, etc. **it is literally just the Go flag package + the ability to have subcommands nicely.
3. It's the same pattern we use in https://github.com/sourcegraph/src-cli (but, well, as a library you can't easily mess up.)

I just figured as long as we're picking a CLI parser I'd throw this one into the hat. I recognize I wrote the library and so am super biased; I don't mind if we go with urfave/cli at all (and, if that's what Keegan or others prefer, we definitely should.)

Feel free to close if you prefer urfave/cli instead! I don't mind :)

## Test plan

* `.bin/sourcegraph --help` works
* `.bin/sourcegraph version` reports the version
* `.bin/sourcegraph` and `.bin/sourcegraph run` start running Sourcegraph
* `sg start app` works

### CODE REVIEW

File: internal/service/svcmain/svcmain.go [hunk 1]
```diff
@@ -29,8 +33,42 @@ type Config struct {
 	AfterConfigure func() // run after all services' Configure hooks are called
 }
 
-// Main is called from the `main` function of the `sourcegraph-oss` and `sourcegraph` commands.
-func Main(services []sgservice.Service, config Config) {
+var commands cmder.Commander
+
+func init() {
```
Comment (slimsag) [inline internal/service/svcmain/svcmain.go:38]: Here's what it looks like to register a new subcommand.

File: internal/service/svcmain/svcmain.go [hunk 2]
```diff
@@ -50,9 +88,32 @@ func Main(services []sgservice.Service, config Config) {
...
+			fmt.Fprintf(flag.CommandLine.Output(), "Usage of 'sourcegraph %s':\n", runFlagSet.Name())
+			runFlagSet.PrintDefaults()
+		},
+	})
+
+	if len(args) == 1 {
+		args = append(args, "run")
+	}
+	golog.SetFlags(0)
+	golog.SetPrefix("")
```
Comment (keegancsmith) [inline internal/service/svcmain/svcmain.go:115]: we are not allowed to use stdlib log, we can add an exemption but that doesn't seem ideal

**Review by keegancsmith — COMMENTED** [review]
your package reminds me a bit of github.com/peterbourgon/ff which I like a lot, although I see it has grown in scope a bit (has support for different config inputs) since I last looked at it. In fact we use ff in zoekt.

To be honest I don't necessarily like urfave, but we use it in other parts of sourcegraph repo (eg the sg tool). I think it is a bit heavyweight, but it feels like the correct engineering decision here is consistency. As the sourcegraph cli may grow in the future, using those extra features (like bash completion) may end up really benefitting us. Especially since we have a vision of shipping one binary so the extra API affordances of urfave may in fact benefit us.

Either way it is a relatively easy thing to change. I am gonna stick with the urfave direction due to the above and the fact the CI is passing on that PR.

**Comment by slimsag** [comment]
sounds good

**Pull request closed without merging**